File: frontend/src/api/alerts.ts
/**Alerts API client (Phase 2 of monitoring plan).

Mirrors the backend's POST/GET/PATCH endpoints under
/api/v1/alerts. Kept in a separate file so the alerts page can
lazy-import only this and the existing types module.
*/
import api from "./client"

export type AlertSeverity = "critical" | "high" | "medium" | "low" | "info"
export type AlertStatus =
  | "new"
  | "acknowledged"
  | "in_progress"
  | "resolved"
  | "false_positive"
export type AlertSource =
  | "wazuh"
  | "elkeid"
  | "syslog"
  | "elastic"
  | "crowdstrike"
  | "sentinelone"
  | "secagent"
  | "unknown"

export interface AlertIOC {
  ips: string[]
  domains: string[]
  hashes: string[]
  urls: string[]
  emails: string[]
  users: string[]
}

export interface AlertRecord {
  alert_id: string
  source: AlertSource | string
  title: string
  description?: string
  severity: AlertSeverity | string
  status: AlertStatus | string
  occurred_at: string
  received_at: string
  hostname?: string
  host_ip?: string
  agent_id?: string
  rule_id?: string
  rule_name?: string
  category?: string
  mitre_attack?: string[]
  iocs?: AlertIOC
  tags?: string[]
  source_url?: string
  raw?: Record<string, unknown>
}

export interface ListAlertsParams {
  severity?: AlertSeverity
  status?: AlertStatus
  source?: AlertSource
  hostname?: string
  limit?: number
  offset?: number
}

export interface ListAlertsResponse {
  items: AlertRecord[]
  limit: number
  offset: number
  count: number
}

export async function listAlerts(params: ListAlertsParams = {}): Promise<ListAlertsResponse> {
  const cleaned: Record<string, string | number> = {}
  if (params.severity) cleaned.severity = params.severity
  if (params.status) cleaned.status = params.status
  if (params.source) cleaned.source = params.source
  if (params.hostname) cleaned.hostname = params.hostname
  if (params.limit !== undefined) cleaned.limit = params.limit
  if (params.offset !== undefined) cleaned.offset = params.offset
  const res = await api.get("/alerts", { params: cleaned })
  return res.data as ListAlertsResponse
}

export async function getAlert(alertId: string): Promise<AlertRecord> {
  const res = await api.get(`/alerts/${encodeURIComponent(alertId)}`)
  return res.data as AlertRecord
}

export async function ingestAlert(
  source: AlertSource,
  payload: Record<string, unknown>
): Promise<{ alert_id: string; received_at: string; severity: AlertSeverity }> {
  const res = await api.post("/alerts/ingest", { source, payload })
  return res.data
}

export async function updateAlertStatus(
  alertId: string,
  status: AlertStatus
): Promise<{ alert_id: string; status: AlertStatus; updated_at: string }> {
  const res = await api.patch(
    `/alerts/${encodeURIComponent(alertId)}/status`,
    { status }
  )
  return res.data
}

File: src/api/routers/alerts.py
"""Alerts router (Phase 1 of monitoring plan).

Endpoints:
  POST /api/v1/alerts/ingest   - accept a vendor-specific JSON, normalize,
                                 persist via AlertStore. This is the primary
                                 inbound path for Wazuh / Elkeid / Syslog
                                 integrations (Phase 1) and the others in
                                 later phases.
  GET  /api/v1/alerts          - list alerts with optional filters.
  GET  /api/v1/alerts/{id}     - single alert.
  PATCH /api/v1/alerts/{id}/status - acknowledge / resolve / false-positive.
"""

from datetime import UTC, datetime

from fastapi import APIRouter, Depends, HTTPException, Query

from src.agents.alert_store import get_alert_store
from src.agents.models import AlertIngestRequest, AlertIngestResponse
from src.api.auth.routes import require_role
from src.common.logging.logger import get_logger
from src.preprocessing.edr_adapter import normalize

logger = get_logger(__name__)

router = APIRouter(prefix="/api/v1/alerts", tags=["alerts"])


@router.post("/ingest", response_model=AlertIngestResponse)
async def ingest_alert(
    body: AlertIngestRequest,
    current_user=Depends(require_role("admin", "analyst")),
) -> AlertIngestResponse:
    """Normalize a vendor JSON into our common Alert model and persist.

    The `source` field in the body is the canonical hint, but if a future
    adapter ships raw bytes (e.g. syslog over webhook) we let the normalizer
    fall back to its default mapping.
    """
    try:
        alert = normalize(body.payload, body.source)
    except Exception as exc:
        logger.warning("alert_normalize_failed", source=str(body.source), error=str(exc))
        raise HTTPException(status_code=400, detail=f"normalize failed: {exc}")

    store = get_alert_store()
    try:
        await store.save_alert(alert)
    except Exception as exc:
        logger.error("alert_persist_failed", alert_id=alert.alert_id, error=str(exc))
        raise HTTPException(status_code=500, detail="persist failed")

    logger.info(
        "alert_ingested",
        alert_id=alert.alert_id,
        source=str(alert.source),
        severity=str(alert.severity),
        hostname=alert.hostname,
    )
    return AlertIngestResponse(
        alert_id=alert.alert_id,
        received_at=alert.received_at,
        severity=alert.severity,
    )


@router.get("")
async def list_alerts(
    severity: str | None = Query(default=None, description="critical/high/medium/low/info"),
    status: str | None = Query(
        default=None, description="new/acknowledged/in_progress/resolved/false_positive"
    ),
    source: str | None = Query(default=None),
    hostname: str | None = Query(default=None),
    limit: int = Query(default=50, ge=1, le=500),
    offset: int = Query(default=0, ge=0),
    current_user=Depends(require_role("admin", "analyst", "viewer")),
) -> dict:
    store = get_alert_store()
    try:
        rows = await store.list_alerts(
            severity=severity,
            status=status,
            source=source,
            hostname=hostname,
            limit=limit,
            offset=offset,
        )
    except Exception as exc:
        raise HTTPException(status_code=500, detail=str(exc))
    return {
        "items": [_serialise(r) for r in rows],
        "limit": limit,
        "offset": offset,
        "count": len(rows),
    }


@router.get("/{alert_id}")
async def get_alert(
    alert_id: str,
    current_user=Depends(require_role("admin", "analyst", "viewer")),
) -> dict:
    store = get_alert_store()
    row = await store.get_alert(alert_id)
    if not row:
        raise HTTPException(status_code=404, detail="alert not found")
    return _serialise(row)


@router.patch("/{alert_id}/status")
async def update_status(
    alert_id: str,
    body: dict,
    current_user=Depends(require_role("admin", "analyst", "responder")),
) -> dict:
    """Update operator workflow state (new -> acknowledged -> in_progress -> resolved)."""
    new_status = body.get("status")
    if not new_status:
        raise HTTPException(status_code=400, detail="status field required")
    valid = ("new", "acknowledged", "in_progress", "resolved", "false_positive")
    if new_status not in valid:
        raise HTTPException(status_code=400, detail=f"status must be one of {valid}")
    store = get_alert_store()
    ok = await store.update_alert_status(alert_id, new_status)
    if not ok:
        raise HTTPException(status_code=404, detail="alert not found")
    return {"alert_id": alert_id, "status": new_status, "updated_at": datetime.now(UTC).isoformat()}


def _serialise(row: dict) -> dict:
    """Convert asyncpg Record (dict) into a JSON-friendly dict.

    asyncpg returns datetime as Python datetime; convert to ISO 8601.
    JSONB columns come back as raw strings (since we serialize on insert);
    parse them back into dicts/list so the UI gets structured data.
    """
    import json

    out = {}
    for k, v in row.items():
        if isinstance(v, datetime):
            out[k] = v.isoformat()
        elif k in ("iocs", "mitre_attack", "tags", "raw") and isinstance(v, str):
            try:
                out[k] = json.loads(v)
            except (TypeError, ValueError):
                out[k] = v
        else:
            out[k] = v
    return out
